### ISSUE TITLE
Looks like quotation cannot be sent to remote

### ISSUE THREAD

**Issue by soloman817** [mbraceproject/MBrace.StarterKit#18]
Hi guys,

I am testing GPU scripting with mbrace cloud scripting. The GPU code is provided through quotation, but I found that in FSI scripting, the quotation cannot be sent to remote side.

here is a simple test of quotation:

```fsharp
let config =
    { Configuration.Default with
        StorageConnectionString = myStorageConnectionString
        ServiceBusConnectionString = myServiceBusConnectionString }

let cluster = Runtime.GetHandle(config)

let str = 
    cloud { return (sprintf "%A" <@ fun a -> a + 1 @>) }
    |> cluster.Run
```

Then I got exception like:

```
System.TypeLoadException: Could not load type '<StartupCode$FSI_0003>.$FSI_0003' from assembly 'FSI-ASSEMBLY_4ddaaa6b-259f-4e7a-9e41-f31323d2ac5a_1, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.
   at FSI_0003.str@27.Invoke(Unit unitVar)
   at MBrace.Builders.Delay@287-1.Invoke(ExecutionContext ctx, Unit t) in c:\Users\eirik\Development\mbrace\MBrace.Core\src\MBrace.Core\Continuation\Builders.fs:line 287
--- End of stack trace from previous location where exception was thrown ---
   在 <StartupCode$MBrace-Azure-Client>.$Process.AwaitResultAsync@135-2.Invoke(Unit _arg4) 位置 c:\Users\eirik\Development\mbrace\MBrace.Azure\src\MBrace.Azure.Client\Process.fs:行号 135
   在 Microsoft.FSharp.Control.AsyncBuilderImpl.args@787-1.Invoke(a a)
   at MBrace.Continuation.ExceptionDispatchInfoUtils.Async.RunSync[T](FSharpAsync`1 workflow, FSharpOption`1 cancellationToken) in c:\Users\eirik\Development\mbrace\MBrace.Core\src\MBrace.Core\Continuation\ExceptionDispatchInfo.fs:line 138
   at <StartupCode$MBrace-Azure-Client>.$Client.RunAsync@232-2.Invoke(Unit unitVar) in c:\Users\eirik\Development\mbrace\MBrace.Azure\src\MBrace.Azure.Client\Client.fs:line 232
   at Microsoft.FSharp.Control.AsyncBuilderImpl.callA@803.Invoke(AsyncParams`1 args)
   at MBrace.Continuation.ExceptionDispatchInfoUtils.Async.RunSync[T](FSharpAsync`1 workflow, FSharpOption`1 cancellationToken) in c:\Users\eirik\Development\mbrace\MBrace.Core\src\MBrace.Core\Continuation\ExceptionDispatchInfo.fs:line 138
   at <StartupCode$FSI_0003>.$FSI_0003.main@() in C:\Users\Xiang\Documents\Projects\GPUCloudDemo\CloudScripts\MBraceTest.fsx:line 26
Stopped due to error
```

**Comment by eiriktsarpalis** [comment]
Hi Xiang,

This is a restriction inherent to the representation of quotation literals in MSIL. I just pushed a commit (https://github.com/nessos/Vagabond/commit/f7481bbf5e46cdc613b2365addd921e9b1df6fba) that somewhat improves the situation, but you can still get errors in certain cases. 

There are two ways this problem can be overcome:
1. Move all quotation literal manipulation code in a class library or
2. Make sure that quotation literals are only used in client-side execution, then passed parametrically to the remote computation. For example, rather than running
```fsharp
let mkQ i = <@ i + 1 @>
cluster.Run (cloud { return mkQ 41 })
```
use the following
```fsharp
let mkQ i = <@ i + 1 @>
let q = mkQ 41
cluster.Run (cloud { return q }) 
```
This should work in every case.

**Comment by dsyme** [comment]
@eiriktsarpalis  - will [the F# 4.0 improvement to IL representation of type references in compiled quotation literals](https://visualfsharp.codeplex.com/SourceControl/network/forks/dsyme/cleanup/contribution/7554) help with this?

**Comment by eiriktsarpalis** [comment]
Possibly, but the error here stems from the opposite pattern: breaking a monolithic dynamic assembly into smaller static slices. Would this case be covered by the changes?

**Comment by dsyme** [comment]
OK, let's dig into this a bit more. Let's look carefully at the F# 3.x static fsc.exe compilation of this:
``
type X() = class end
let f () = <@ fun a -> a + 1, X() @>
``
We see these relevant parts:
``
  IL_0000:  ldtoken    '<StartupCode$a>'.$A
...
  IL_001f:  ldtoken    field valuetype '<PrivateImplementationDetails$a>'/T1598_503Bytes@ A::field1599@
...
  IL_0029:  call       class [FSharp.Core]Microsoft.FSharp.Quotations.FSharpExpr [FSharp.Core]Microsoft.FSharp.Quotations.FSharpExpr::Deserialize(...)
...
``
*Problem 1*. The first problem for slicing is this line:
``
  IL_0000:  ldtoken    '<StartupCode$a>'.$A
``
In F# 3.x this is fetching a type token to represent "the current assembly" to pass to [Deserialize](https://github.com/fsharp/fsharp/blob/master/src/fsharp/FSharp.Core/quotations.fsi#L334).  In F# 3.x quotation literal bytes, references to types in the current assembly like that implied by the constructor call "X()" are encoded as ".", see [here](https://github.com/fsharp/fsharp/blob/master/src/fsharp/FSharp.Core/quotations.fs#L1249). These references are resolved via the purely dummy type passed in here. 

When you slice F# 3.1 code into multiple assemblies, these intra-assembly references implied by the quotation bytes will no longer be valid.  While you can pass in any suitable type from "the current assembly" as this qualifying type, that won't be sufficient to resolve intra-assembly type references that cross slice boundaries like "X" in the example above.  This is hard to workaround short of manually rewriting the F# 3.x quotation literal bytes so that intra-assembly references become cross-assembly references. I think it would be feasible for Vagrant to do this rewriting but is quite a lot of work.

From F# 4.0 (if referencing FSharp.Core 4.4.0.0+), both inter-assembly and intra-assembly type references are encoded as integer references into a table of System.Type objects passed into deserialization.   You still see this (still passing in a "dummy type from this assembly" though that's ignored AFAIK)
``
  IL_0000:  ldtoken    '<StartupCode$a>'.$A
``
but you also see this:
``
  IL_0034:  ldtoken    A/X
``
plus quite a lot of similar ldtoken/typeof calls for mscorlib and FSharp.Core types used by the quotation. My understanding is that this will now slice correctly. 

*Problem 2*. The second source of trouble is this (which is present in both F# 3.1 and 4.0):
``
  IL_001f:  ldtoken    field valuetype '<PrivateImplementationDetails$a>'/T1598_503Bytes@ A::field1599@
``
This is loading a reference to the "field in a fake struct type" used in the .,NET encoding of static array data.  For a quotation literal this type and its corresponding field data should be brought into the slice where the Deserialize call is located. Beware that these types are memoized in ilxgen.fs AssemblyBuilder so they may get copied into more than one slice if another array turns up that has exactly the same size as a previious array of bytes.  I think your fix is enough to make sure this works ok.

Your fix seems to be enough to fix *Problem 2*, and also to make F# 3.1 quotation literals that don't have intra-assembly type references (i.e. references to types defined in the F# interactive session, and only refer to types from other DLLs) work OK, since they don't hit *Problem 1*.  

I think the F# 4.0 change will be enough to also fix *Problem 1*, though your fix to Vagrant will still be needed for *Problem 2*.

Sorry for the long explanation, I'll edit and update this as we get more info.

**Comment by eiriktsarpalis** [comment]
Thanks for the detailed explanation. I haven't studied the code that F# 4.0 generates in this case, but assuming that the `referencedTypeDefs` are contained in IL and not binary form, then everything should work fine; Vagabond ensures that all typerefs are correctly remapped in generated slices.

**Comment by eiriktsarpalis** [comment]
BTW, I'd be interested to know what led you to use pickling for loading quotations.

**Comment by soloman817** [comment]
@eiriktsarpalis I'm coding a demo to see how to integrate Alea GPU with mbrace cloud. I think there are some reasons for sending quotations to cloud:

In Alea GPU, if we work with F#, we use quotation to represent the GPU code. although Alea GPU support IL decodidng, but that works with C#, and for F#, using quotation is more natural way and is more powerful.

So, the GPU source code is represented through quotations, then we have two ways to compile it:

```fsharp
let gpuModuleSource = cuda {
    let! gpuKernelFunction = <@ somefun @> |> Compiler.DefineKernel
    return Entry (blablabla) }

let result = cloud {
    use program = GPUWorker.Default.LoadProgram(gpuModuleSource)
    return program.Run blablabla }
```

In the above code, the gpu code is send to cloud and then JIT compiled and then loaded into the gpu worker (we call it also worker). and this need to send qutoations to cloud, because the gpu kernel function is represented via quotation.

we can also compile gpu module locally and send the module to cloud like:

```fsharp
let gpuModuleSource = cuda {
    let! gpuKernelFunction = <@ somefun @> |> Compiler.DefineKernel
    return Entry (blablabla) }

let gpuModule = gpuModuleSource |> Compiler.compile |> Compiler.link

let result = cloud {
    use program = GPUWorker.Default.LoadProgram(gpuModule)
    return program.Run blablabla }
```

It has two issues of the second way.

1. you need explicitly compile the gpu source code. Also, if we do the JIT remotely, we can compile this source code with more information, such as the GPU's arch, etc. and results better gpu binary. of course you can get those information through cloud computation, but that results many code, not good for user.

2. alea gpu requires compile license, so for a cloud, it would be good to just deploy the license on the cloud worker, then on your client, you don't need to have a license, because you don't need to compile it locally.

BTW, I will soon finish a demo with alea gpu + mbrace cloud demo. and in that demo, I will code the gpu part in a normal F# assembly, and reference it in the fsx file, that works.

Regards,
Xiang.

**Comment by eiriktsarpalis** [comment]
@dsyme I can confirm that the changes to quotation literal representation introduced in F# 4.0 have fixed any Vagabond-related problems.

**Comment by dsyme** [comment]
Thank you, that's great to know!

**Comment by dsyme** [comment]
Can we close this?  @soloman817 do you want to try again with the GPU tests using F# 4.0?

**Comment by soloman817** [comment]
Good to know it is fixed with F# 4.0, so I close it, and will find some time to test that with F# 4.0.
